Remove `i128` and `u128` from `improper_ctypes_definitions`

tgross35 opened the pull request:
Rust's 128-bit integers have historically been incompatible with C [1]. However, there have been a number of changes in Rust and LLVM that mean this is no longer the case:

* Incorrect alignment of `i128` on x86 [1]: adjusting Rust's alignment proposed at https://github.com/rust-lang/compiler-team/issues/683, implemented at https://github.com/rust-lang/rust/pull/116672.
* LLVM version of the above: resolved in LLVM, including ABI fix. Present in LLVM18 (our minimum supported version).
* Incorrect alignment of `i128` on 64-bit PowerPC, SPARC, and MIPS [2]: Rust's data layouts adjusted at https://github.com/rust-lang/rust/pull/132422, https://github.com/rust-lang/rust/pull/132741, https://github.com/rust-lang/rust/pull/134115.
* LLVM version of the above: done in LLVM 20 https://github.com/llvm/llvm-project/issues/102783.
* Incorrect return convention of `i128` on Windows: adjusted to match GCC and Clang at https://github.com/rust-lang/rust/pull/134290.

At https://github.com/rust-lang/lang-team/issues/255#issuecomment-2088855084, the lang team considered it acceptable to remove `i128` from `improper_ctypes_definitions` if the LLVM version is known to be compatible. Time has elapsed since then and we have dropped support for LLVM versions that do not have the x86 fixes, meaning a per-llvm-version lint should no longer be necessary. The PowerPC, SPARC, and MIPS changes only came in LLVM 20 but since Rust's datalayouts have also been updated to match, we will be using the correct alignment regardless of LLVM version.

`repr(i128)` was added to this lint in https://github.com/rust-lang/rust/pull/138282, but is also removed here.

Part of the decision is that `i128` should match `__int128` in C on platforms that provide it, which documentation is updated to indicate. We will not guarantee that `i128` matches `_BitInt(128)` since that can be different from `__int128`. Some platforms (usually 32-bit) do not provide `__int128`; if any ABIs are extended in the future to define it, we will need to make sure that our ABI matches.

Closes: https://github.com/rust-lang/rust/issues/134288

[1]: https://github.com/rust-lang/rust/issues/54341
[2]: https://github.com/rust-lang/rust/issues/128950

tgross35 committed:
Remove `i128` and `u128` from `improper_ctypes_definitions`

Rust's 128-bit integers have historically been incompatible with C [1].
However, there have been a number of changes in Rust and LLVM that
mean this is no longer the case:

* Incorrect alignment of `i128` on x86 [1]: adjusting Rust's alignment
  proposed at https://github.com/rust-lang/compiler-team/issues/683,
  implemented at https://github.com/rust-lang/rust/pull/116672.
* LLVM version of the above: resolved in LLVM, including ABI fix.
  Present in LLVM18 (our minimum supported version).
* Incorrect alignment of `i128` on 64-bit PowerPC, SPARC, and MIPS [2]:
  Rust's data layouts adjusted at
  https://github.com/rust-lang/rust/pull/132422,
  https://github.com/rust-lang/rust/pull/132741,
  https://github.com/rust-lang/rust/pull/134115.
* LLVM version of the above: done in LLVM 20
  https://github.com/llvm/llvm-project/issues/102783.
* Incorrect return convention of `i128` on Windows: adjusted to match
  GCC and Clang at https://github.com/rust-lang/rust/pull/134290.

At [3], the lang team considered it acceptable to remove `i128` from
`improper_ctypes_definitions` if the LLVM version is known to be
compatible. Time has elapsed since then and we have dropped support for
LLVM versions that do not have the x86 fixes, meaning a per-llvm-version
lint should no longer be necessary. The PowerPC, SPARC, and MIPS changes
only came in LLVM 20 but since Rust's datalayouts have also been updated
to match, we will be using the correct alignment regardless of LLVM
version.

`repr(i128)` was added to this lint in [4], but is also removed here.

Part of the decision is that `i128` should match `__int128` in C on
platforms that provide it, which documentation is updated to indicate.
We will not guarantee that `i128` matches `_BitInt(128)` since that can
be different from `__int128`. Some platforms (usually 32-bit) do not
provide `__int128`; if any ABIs are extended in the future to define it,
we will need to make sure that our ABI matches.

Closes: https://github.com/rust-lang/rust/issues/134288
Closes: https://github.com/rust-lang/rust/issues/128950

[1]: https://github.com/rust-lang/rust/issues/54341
[2]: https://github.com/rust-lang/rust/issues/128950
[3]: https://github.com/rust-lang/lang-team/issues/255#issuecomment-2088855084
[4]: https://github.com/rust-lang/rust/pull/138282

tgross35 commented:
I am not aware of any other incompatibilities at this time, @beetrees is there anything outstanding?

I believe this needs a lang FCP @rust-lang/lang.

@rustbot label +T-lang, +I-lang-nominated, +I-lang-easy-decision

RalfJung commented:
Cc @workingjubilee

programmerjake commented:
> `_BitInt(128)` currently has a different alignment than `__int128` on x86-64: it would be good to document somewhere which one Rust's `i128` is compatible with.

ideally x86-64 would fix its broken ABIs so `__int128` and `_BitInt(128)` have the same ABI.

bjorn3 commented:
I would expect i128 arguments that are passed on the stack to be aligned to 16 bytes with the fix and to be aligned to 4 or 8 bytes without the fix. This changes the offset where the argument is passed and thus affects the ABI.

nikic commented:
> While 32-bit platforms don't support `__int128`, they do support `_BitInt(128)`, so if the lint is going to be removed on all platforms (not just 64-bit platforms) then I'd expect `i128` and `_BitInt(128)` to be ABI compatible on 32-bit platforms. Spot checking a few targets now, I've discovered:

Given that the fact that i128 maps to __int128, but not _BitInt(128) on x86_64, I don't think that mapping it to _BitInt(128) on other platforms is a good idea.

I think it would be preferable to treat i128 as not a proper FFI type on platforms that do not actually specify an ABI for __int128.

petrochenkov commented:
r? @workingjubilee

petrochenkov commented:
r? @RalfJung

RalfJung commented:
Sorry, I don't have the low level ABI / LLVM knowledge to review this.

r? @nikic

workingjubilee commented:
I don't actually think this will be an easy decision.

workingjubilee commented:
This FIXME does not seem to be addressed for 64-bit Power: https://github.com/rust-lang/rust/blob/a18bd8acfc32dbfbbe150cd52eecf24e67fb69b6/compiler/rustc_target/src/callconv/powerpc64.rs#L1-L3

RalfJung commented:
Also see [Zulip](https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202025-02-20/near/500912724)[#t-compiler/meetings > &#91;weekly&#93; 2025-02-20 @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202025-02-20/near/500912724)

RalfJung commented:
Is there precedent for improper_ctypes being target-dependent? That seems problematic as the crate author seems likely to not see the diagnostic and users will not see it either when it occurs in a dependency.

bors commented:
:umbrella: The latest upstream changes (presumably #140043) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

workingjubilee commented:
> > Is there precedent for improper_ctypes being target-dependent? That seems problematic as the crate author seems likely to not see the diagnostic and users will not see it either when it occurs in a dependency.
> 
> AFAIK no: the closest would be AIX having a lint for when Rust's `repr(C)` struct layout doesn't match the C compiler's, but that's done via a separate `uses_power_alignment` lint (added in #135552).

note for posterity: this lint was found to be a pile of confusion that wasn't really addressing the right problem. I do not regret merging the lint because meandering down the confused path, despite proving slightly aimless, eventually surfaced the real problems. however, it should probably be considered an example to avoid following.

I do not think this fact should necessarily change anyone's decisions here, just making clear that it was using linting Rust source to solve a problem so deep it was instead about the compiler's model of the target. that problem lives in either rustc's backend crates or LLVM, depending on your perspective.

RalfJung commented:
>  it's not the definitive reference on what Rust considers to be ABI compatible with C.

It is, however, the *only* reference that exists for this at the moment, to my knowledge...

tgross35 commented:
To be clear; this PR removes the lint on all targets and does not add a check against a list of known good/problematic targets. Is that what is desired, or is there still interest in the target-specific lint?

I think it would be a good idea to update the `i128` docs to say our `i128` is meant to be the same as `__int128` on platforms where that exists, but not necessarily with `_BitInt(128)`, which seems to be consensus here. If that sounds reasonable, I'll make that change as well.

tgross35 commented:
Rebased. I also removed the lint against `repr(i128)` enums that fell into this group and was added in https://github.com/rust-lang/rust/pull/138282.

Also added a note about ABI compatibility to the primitive docs.

@rustbot ready

bors commented:
:umbrella: The latest upstream changes (presumably #141717) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tgross35 commented:
Rebased. @workingjubilee I think this mostly just needs an ack from you regarding the new documentation; the rest is pretty straightforward.

workingjubilee commented:
Thank you, apologies for the latency. This looks good to ship.

@bors r+ rollup

bors commented:
:pushpin: Commit 0cba7fb6f665300ba0febeff18f34cc74ff44224 has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 0cba7fb6f665300ba0febeff18f34cc74ff44224 -->
<!-- homu: {"type":"Approved","sha":"0cba7fb6f665300ba0febeff18f34cc74ff44224","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented:
@bors r=traviscross,workingjubilee

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #141229

bors commented:
:pushpin: Commit 0cba7fb6f665300ba0febeff18f34cc74ff44224 has been approved by `traviscross,workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=traviscross,workingjubilee 0cba7fb6f665300ba0febeff18f34cc74ff44224 -->
<!-- homu: {"type":"Approved","sha":"0cba7fb6f665300ba0febeff18f34cc74ff44224","approver":"traviscross,workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
